crates/nu-command/src/conversions/into/record.rs

```diff
@@ -103,45 +100,61 @@ impl Command for SubCommand {
-    let res = match input {
-        Value::Date { val, .. } => parse_date_into_record(val, span),
-        Value::Duration { val, .. } => parse_duration_into_record(val, span),
-        Value::List { mut vals, .. } => match input_type {
-            Type::Table(..) if vals.len() == 1 => vals.pop().expect("already checked 1 item"),
-            _ => Value::record(
-                vals.into_iter()
-                    .enumerate()
-                    .map(|(idx, val)| (format!("{idx}"), val))
-                    .collect(),
-                span,
-            ),
-        },
-        Value::Range { val, .. } => Value::record(
-            val.into_range_iter(span, engine_state.signals().clone())
-                .enumerate()
-                .map(|(idx, val)| (format!("{idx}"), val))
-                .collect(),
+fn into_record(call: &Call, input: PipelineData) -> Result<PipelineData, ShellError> {
+    let span = input.span().unwrap_or(call.head);
+    match input {
+        PipelineData::Value(Value::Date { val, .. }, _) => {
+            Ok(parse_date_into_record(val, span).into_pipeline_data())
+        }
+        PipelineData::Value(Value::Duration { val, .. }, _) => {
+            Ok(parse_duration_into_record(val, span).into_pipeline_data())
+        }
+        PipelineData::Value(Value::List { .. }, _) | PipelineData::ListStream(..) => {
+            let mut record = Record::new();
+            let metadata = input.metadata();
+            for item in input.into_iter() {
+                // Don't use .extend() unless that gets changed to check for duplicate keys
+                let span = item.span();
+                match item {
+                    Value::Record { val, .. } => {
+                        for (key, val) in val.into_owned() {
+                            record.insert(key, val);
+                        }
+                    }
+                    Value::List { mut vals, .. } => {
+                        if vals.len() == 2 {
+                            let (val, key) = vals.pop().zip(vals.pop()).expect("length is < 2");
+                            record.insert(key.coerce_into_string()?, val);
+                        } else {
+                            return Err(ShellError::IncorrectValue {
+                                msg: format!(
+                                    "expected list with two elements, but found {}",
+                                    vals.len()
+                                ),
+                                val_span: span,
+                                call_span: call.head,
+                            });
+                        }
+                    }
+                    Value::Nothing { .. } => {}
+                    Value::Error { error, .. } => return Err(*error),
+                    _ => {
+                        return Err(ShellError::TypeMismatch {
+                            err_message: "expected record or list with two elements".into(),
```

Comment:
Putting aside my deep hatred for the current `TypeMismatch` variant, maybe also worth being more explicit.

```suggestion
                            err_message: "To create a record from a list, expected items of type record or inner lists with two elements".into(),
```